source/slang/slang-capability.cpp

```diff
@@ -134,1104 +140,618 @@ bool isCapabilityDerivedFrom(CapabilityAtom atom, CapabilityAtom base)
+        thisTargetSet.target = otherTargetSet.first;
+        thisTargetSet.shaderStageSets.reserve(otherTargetSet.second.shaderStageSets.getCount());
+        thisTargetSet.unionWith(otherTargetSet.second);
+    }
+}
 
-    // Next, we want to compare the candidates to the `targetCaps`
-    // to figure out whether one is obviously "more specialized" for
-    // the target.
-    //
-    // We measure the degree to which a candidate is specialized for
-    // the target as the size of its set intersection with `targetCaps`.
-    //
-    // TODO: If both `candidateCaps` and `existingCaps` are implied
-    // by `targetCaps`, then this amounts to just measuring the
-    // size of each set. We probably want this size-based check to
-    // come later in the overall process.
-    //
-    // TODO: A better model here might be to actually compute the actual
-    // intersected sets, and then check if one is a super-set of the other.
-    //
-    auto candidateIntersectionSize = targetCaps.countIntersectionWith(candidateCaps);
-    auto existingIntersectionSize = targetCaps.countIntersectionWith(existingCaps);
-    if(candidateIntersectionSize != existingIntersectionSize)
-        return candidateIntersectionSize > existingIntersectionSize;
-
-    // Next we want to consider that if one of the two candidates
-    // is actually available on the target (meaning that it is
-    // implied by `targetCaps`) then we probably want to pick that one
-    // (since we can use that candidate on the chosen target without
-    // enabling any additional features the user didn't ask for).
-    //
-    // TODO: This step currently needs to come after the preceeding
-    // one because otherwise we risk selecting a `__target_intrinsic`
-    // decoration with *no* requirements (which are currently being
-    // added implicitly in many places) over any one with explicit
-    // requirements (since every target implies the empty set of
-    // requirements).
-    //
-    // In many ways the counting-based logic above amounts to a quick
-    // fix to prefer a non-empty set of requirements over an empty one,
-    // so long as something in that non-empty set overlaps with the target.
-    //
-    // TODO: The best fix is probably to figure out how "catch-all"
-    // intrinsic function definitions should be encoded; we clearly
-    // want them to be used only as a fallback when no target-specific
-    // variants are present.
-    //
-    bool candidateIsAvailable = targetCaps.implies(candidateCaps);
-    bool existingIsAvailable = targetCaps.implies(existingCaps);
-    if(candidateIsAvailable != existingIsAvailable)
-        return candidateIsAvailable;
+/// Join sets, but: 
+/// 1. do not destroy target set's which are incompatible with `other` (destroying shaderStageSets is fine)
+/// 2. do not create an `CapabilityAtom::Invalid` target set.
+void CapabilitySet::nonDestructiveJoin(const CapabilitySet& other)
```

Comment:
prototype is in the header, we do this for compile speeds I thought (change 1 compilation unit versus all that include the header)?

Reply:
What compile speed are you talking about?
Do you mean that if we write comments on the header files, it will increase the compile time?

Your comment here is something called "contract". You are making promises of how the function will behave.  And it is expected to read by the API users. And API users often don't have access to the source file because when the library is sold for money, it doesn't come with cpp files.
For the reason, contracts are written on the header files where the API users can see.

FYI, when you write a contract, there are rules you need to follow. Two main things are "requirements", and "effect". When the requirement is not met, the promised effect will not happen. You need to make it clear on what the effect will be when the requirement was met.

Reply:
Another thing that is related to this is "substitution principle". I hope we can have a chance to talk about it soon but not this time.

Reply:
I misread. The comment should be in the header.

Reply:
hmm, I already had a comment on the header...

Reply:
and the comment is correct, huh.